### PR TITLE
optimize `abstract_call_gf_by_type`

### PR DESCRIPTION
**Pull request by aviatesk**
Might be better to merge after JuliaLang/julia#56565.

@nanosoldier `runbenchmarks("inference", vs=":master")`

### CODE REVIEW

File: Compiler/src/abstractinterpretation.jl [hunk 1]
```diff
@@ -50,35 +58,34 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
...
-    const_results = nothing # or const_results::Vector{Union{Nothing,ConstResult}} if any const results are available
-    fargs = arginfo.fargs
-    all_effects = EFFECTS_TOTAL
-    slotrefinements = nothing # keeps refinement information on slot types obtained from call signature
+    rettype = SafeBox{Any}(Bottom)
+    exctype = SafeBox{Any}(Bottom)
+    conditionals_boxed = SafeBox{Union{Nothing,Tuple{Vector{Any},Vector{Any}}}}(nothing) # keeps refinement information of call argument types when the return type is boolean
+    seenall = SafeBox{Bool}(true)
+    const_results_boxed = SafeBox{Union{Nothing,Vector{Union{Nothing,ConstResult}}}}(nothing) # if any const results are available
+    all_effects = SafeBox{Effects}(EFFECTS_TOTAL)
```
Comment (vtjnash) [inline Compiler/src/abstractinterpretation.jl:73]: A single mutable struct might be more efficient here than the 7 separate allocations, if we want to work on trying to fine tune this

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56572) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/5eb1a7e_vs_5ec3215/report.md).

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56572#issuecomment-2480421394) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/b35a596_vs_caa2f7d/report.md).

File: Compiler/src/abstractinterpretation.jl (hunk 1 above)
Comment (aviatesk) [reply Compiler/src/abstractinterpretation.jl]: Right. I’ve updated the implementation accordingly.

**Pull request merged**